fuse/file.go

```diff
@@ -3,103 +3,370 @@ package fuse
 	}
+
+	activity.Record(ctx, "FUSE: Read %v/%v bytes starting at %v from %v", len(resp.Data), req.Size, req.Offset, f)
 	return nil
 }
 
-func (fh fileHandle) Read(ctx context.Context, req *fuse.ReadRequest, resp *fuse.ReadResponse) error {
-	data, err := plugin.ReadWithAnalytics(ctx, fh.r, int64(req.Size), req.Offset)
-	if err == io.EOF {
-		// If we don't ignore this, then cat will display an input/output error message
-		// for entries with unknown content size.
-		err = nil
+var _ = fs.HandleWriter(&file{})
+
+func (f *file) Write(ctx context.Context, req *fuse.WriteRequest, resp *fuse.WriteResponse) error {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+
+	// Ensure handle is in list of writers.
+	f.writers[req.Handle] = struct{}{}
+
+	if f.isFileLike() {
+		// If starting write beyond the current length, read to fill it in.
+		if start := int64(len(f.data)); req.Offset > start {
+			data, err := f.load(ctx, start, req.Offset)
+			if err != nil {
+				return err
+			}
+			f.data = append(f.data, data...)
+		}
+	}
+
+	// Expand the buffer if necessary to store the write data.
+	newLen := req.Offset + int64(len(req.Data))
+	if newLen := int(newLen); newLen > len(f.data) {
+		f.data = append(f.data, make([]byte, newLen-len(f.data))...)
+	}
+
+	// If file-like, then update readable size to reflect the expanded buffer.
+	if f.isFileLike() && f.readSize < uint64(newLen) {
+		f.readSize = uint64(newLen)
+	}
+
+	resp.Size = copy(f.data[req.Offset:], req.Data)
+	activity.Record(ctx, "FUSE: Write %v/%v bytes starting at %v from %v", resp.Size, len(req.Data), req.Offset, f)
+	return nil
+}
+
+func (f *file) load(ctx context.Context, start, end int64) ([]byte, error) {
+	if !plugin.ReadAction().IsSupportedOn(f.entry) {
+		activity.Warnf(ctx, "FUSE: Non-contiguous writes (at %v) unsupported on %v", start, f)
+		return nil, fuse.ENOTSUP
+	}
+
+	return plugin.Read(ctx, f.entry, end-start, start)
+}
+
+var _ = fs.HandleFlusher(&file{})
+
+// Note that this implementation of Flush only calls plugin.Write if there were previous calls to
+// Write or Setattr. It doesn't check whether the data that's there matches what we're writing.
+func (f *file) Flush(ctx context.Context, req *fuse.FlushRequest) error {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+	activity.Record(ctx, "FUSE: Flush %v: %+v", f, *req)
+
+	if _, ok := f.writers[req.Handle]; !ok {
+		return nil
+	}
+
+	// If this handle had an open writer, write current data.
+	dataLen := int64(len(f.data))
+	if f.isFileLike() {
+		// Only file-like entries keep data and readSize in sync.
+		if uint64(dataLen) > f.readSize {
+			panic("Size was not kept up-to-date with changes to data.")
+		}
+
+		if uint64(dataLen) < f.readSize {
+			// Missing some data, load the remainder before writing.
+			data, err := f.load(ctx, dataLen, int64(f.readSize))
+			if err != nil && err != io.EOF {
+				return err
+			}
+			f.data = append(f.data, data...)
+
+			// If still too small then something increased the size beyond the original.
```

Comment:
When/how could this happen? Should this be a known issue/something we document?

Reply:
Would've been a call to Setattr. Rewriting this comment.